constraints/src/test/java/gov/nasa/jpl/aerie/constraints/time/WindowsTest.java

```diff
@@ -1,559 +1,506 @@
+                               Segment.of(interval(7, 8, SECONDS), false),
+                               Segment.of(interval(8, Exclusive, 10, Exclusive, SECONDS), true),
+                               Segment.of(interval(13, 14, SECONDS), true),
+                               Segment.of(interval(14, Exclusive, 16, Exclusive, SECONDS), false),
+                               Segment.of(at(Duration.MAX_VALUE.minus(Duration.of(1, SECONDS))), true)); //long overflow if at max value
 
-    windows.intersectWith(Window.EMPTY);
+    Windows result = orig.shiftBy(Duration.of(-1, SECONDS), Duration.of(1, SECONDS));
 
-    final var expected = new Windows();
 
-    assertEquivalent(expected, windows);
+    Windows expected = new Windows(Segment.of(interval(-1, 1, SECONDS), true),
+                                   Segment.of(interval(1, Exclusive, 2, Inclusive, SECONDS), false),
+                                   Segment.of(interval(4, Inclusive, 11, Exclusive, SECONDS), true),
+                                   Segment.of(interval(12, 15, SECONDS), true),
+                                   Segment.of(interval(15, Exclusive, 16, Exclusive, SECONDS), false),
+                                   Segment.of(interval(Duration.MAX_VALUE.minus(Duration.of(2, SECONDS)),
+                                                       Duration.MAX_VALUE), true));
+    assertIterableEquals(expected, result);
   }
 
   @Test
-  public void intersectAdjacent() {
-    final var windows = new Windows();
-    windows.add(window(0,  1, MICROSECONDS));
-    windows.add(window(2,  4, MICROSECONDS));
+  public void shiftByFromStartFromEndPermsWithInterval() {
+    var leftEnd = interval(0, 2, SECONDS);
+    var rightEnd = interval(8, 10, SECONDS);
 
-    windows.intersectWith(1,  2, MICROSECONDS);
+    var orig = new Windows()
+        .set(new Windows(Segment.of(leftEnd, true), Segment.of(rightEnd, true)));
 
-    final var expected = new Windows();
-    expected.addPoint(1, MICROSECONDS);
-    expected.addPoint(2, MICROSECONDS);
+    var fromStartPosFromEndPos = orig.shiftBy(Duration.of(-1, SECONDS), Duration.of(1, SECONDS));
+    assertIterableEquals(fromStartPosFromEndPos, new Windows(
+        Segment.of(interval(-1, 3, SECONDS), true),
+        Segment.of(interval(7, 11, SECONDS), true)
+    ));
 
-    assertEquivalent(expected, windows);
-  }
+    var fromStartPosFromEndNeg = orig.shiftBy(Duration.of(1, SECONDS), Duration.of(-1, SECONDS));
+    assertEquals(fromStartPosFromEndNeg, new Windows(
+        Segment.of(interval(0, Inclusive, 1, Exclusive, SECONDS), false),
+        Segment.of(interval(1, 1, SECONDS), true),
+        Segment.of(interval(1, Exclusive, 2, Inclusive, SECONDS), false),
+        Segment.of(interval(8, Inclusive, 9, Exclusive, SECONDS), false),
+        Segment.of(interval(9, 9, SECONDS), true),
+        Segment.of(interval(9, Exclusive, 10, Inclusive, SECONDS), false)
+    ));
```

Comment:
The `assertEquals` method takes `expected` first and `actual` second, so this will produce a misleading error message when the assertion fails.

There are a lot of these. Sorry.

Reply:
Yes I saw them too, but I decided it wasn't worth the time it takes to fix them.